fuel-vm/src/interpreter/crypto.rs

```diff
@@ -202,3 +246,219 @@ pub(crate) fn sha256(
+        crate::error::PanicOrBug::Panic(fuel_tx::PanicReason::InvalidEllipticCurvePoint)
+    })?;
+    let bx = Fq::from_slice(
+        memory.read(
+            point_ptr
+                .checked_add(96)
+                .ok_or(crate::error::PanicOrBug::Panic(
+                    fuel_tx::PanicReason::ArithmeticOverflow,
+                ))?,
+            32u64,
+        )?,
+    )
+    .map_err(|_| {
+        crate::error::PanicOrBug::Panic(fuel_tx::PanicReason::InvalidEllipticCurvePoint)
+    })?;
+    let a = Fq2::new(ax, ay);
+    let b = Fq2::new(bx, by);
+    if a.is_zero() && b.is_zero() {
+        Ok(G2::zero())
+    } else {
+        Ok(G2::from(AffineG2::new(a, b).map_err(|_| {
+            crate::error::PanicOrBug::Panic(
+                fuel_tx::PanicReason::InvalidEllipticCurvePoint,
+            )
+        })?))
+    }
+}
+
+pub(crate) fn ec_operation(
+    memory: &mut MemoryInstance,
+    owner: OwnershipRegisters,
+    pc: RegMut<PC>,
+    dst: Word,
+    curve_id: Word,
+    operation_type: Word,
+    points_ptr: Word,
+) -> SimpleResult<()> {
+    match curve_id {
+        0 => {
+            match operation_type {
+                // Two points addition
+                0 => {
+                    let point1 = read_g1_point_alt_bn_128(memory, points_ptr)?;
+                    let point2 = read_g1_point_alt_bn_128(
+                        memory,
+                        points_ptr.checked_add(64).ok_or(
+                            crate::error::PanicOrBug::Panic(
+                                fuel_tx::PanicReason::ArithmeticOverflow,
+                            ),
+                        )?,
+                    )?;
+                    let mut output = [0u8; 64];
+                    #[allow(clippy::arithmetic_side_effects)]
+                    if let Some(sum) = AffineG1::from_jacobian(point1 + point2) {
+                        sum.x().to_big_endian(&mut output[..32]).unwrap();
+                        sum.y().to_big_endian(&mut output[32..]).unwrap();
+                    }
+                    memory.write_bytes(owner, dst, output)?;
+                }
+                // Scalar multiplication
+                1 => {
+                    let point = read_g1_point_alt_bn_128(memory, points_ptr)?;
+                    let scalar = Fr::from_slice(memory.read(
+                        points_ptr.checked_add(64).ok_or(
+                            crate::error::PanicOrBug::Panic(
+                                fuel_tx::PanicReason::ArithmeticOverflow,
+                            ),
+                        )?,
+                        32u64,
+                    )?)
+                    .map_err(|_| {
+                        crate::error::PanicOrBug::Panic(
+                            fuel_tx::PanicReason::InvalidEllipticCurvePoint,
+                        )
+                    })?;
+                    let mut output = [0u8; 64];
+                    #[allow(clippy::arithmetic_side_effects)]
```

Comment:
How do we know the multiplication cannot panic? There should be a `// Safety: ...` comment to explain that.

Reply:
Added a comment.